heft-plugins/heft-storybook-plugin/src/schemas/storybook.schema.json

```diff
@@ -28,6 +28,11 @@
       "description": "If this is empty, then it will use the storybook default output dir. If you want to change the static build output dir to staticBuildDir, then the static build output dir would be: `\"staticBuildOutputFolder\": \"staticBuildDir\"`",
       "type": "string",
       "pattern": "[^\\\\]"
+    },
+    "storybookPackageNameTarget": {
+      "title": "Specifies an NPM dependency name that is used as the (cwd) target for the storybook commands",
+      "description": "By default the plugin executes the storybook commands in the local package context, but for distribution purposes it can be useful to split the TS library and storybook exports into two packages. For example, If you create an 'empty' storybook 'app' project for distribution purposes, then the storybook package name would be: `\"storybookPackageName\": \"my-ui-storybook-library\"`",
+      "type": "string"
```

Comment:
Because this is an already working solution for people who need to use Storybook 7, I think we should accept it.

However in the [Zulip thread](https://rushstack.zulipchat.com/#narrow/stream/262522-heft/topic/Storybook.20v7/near/361429415) I suggested a larger change:

> It's been a while, but my guess is that the `startupModulePath` was designed this way because packages like `@storybook/react` had their own CLI binaries, and we were relying on internal script paths because there was no automation API. If it changed, the Heft user could fix it in the config file, but of course it's annoying for the Heft user to have to know how to find the `bin/build.js` path.
> 
> 1. We should double-check whether Storybook still has neglected to provide an automation API.  (Is the CLI the only way to invoke these commands? Maybe they improved that too.)
> 
> 2. Since the `startupModulePath` underlying system is completely changed in V7, I think we could eliminate the need for those settings. (I suppose we could repurpose them to allow custom scripts like your hack above.)
> 
> Looking closer at Storybook 7:
> 
> - The `@storybook/cli` package creates the `sb` and `storybook` shell binaries that call the real **index.js**
> - The `@storybook/cli` contains the actual business logic; I don't think there's any automation API.
> - The `storybook`  package is a "wrapper" that also creates conflicting `sb` and `storybook` binaries whose **index.js** does `require('@storybook/cli/bin/index');`.  This is a bad practice that will produce warnings when the package manager isn't sure which binary to create, but okay.
> 
> So from the perspective of Heft who is trying to find `sb` and invoke it, we still have all the same problems: 
> 
> 1. To ensure deterministic results we need to probe for NPM dependencies, not shell commands in an unreliable `PATH` or `node_modules/.bin/` folder (where the Storybook maintainers are overwriting their own script names).  
> 
> 2. And once we find the CLI we should examine the associated **package.json** to determine which CLI parameters are supported. But which package name? `@storybook/react`? `@storybook/cli`? `storybook`?
> 
> 3. In a monorepo, there's also a concern about [phantom dependencies](https://rushjs.io/pages/advanced/phantom_deps/). We need to be careful about calling `require("@storybook/cli")` for example in a folder whose **package.json** doesn't actually depend on `@storybook/cli`.
>
> ### How about a new design like this?
> 
> ```
> /**
>  * Specify the NPM package that provides the CLI binary to run.
>  * It will be resolved from the folder of your storykit package.
>  */
> // other possible values:  @storybook/react, storybook, etc.
> "cliPackageName": "@storybook/cli",
> 
> /** 
>  * Specify how the Storybook CLI should be invoked.  Possible values:
>  *
>  *  - "storybook6": For a static build, Heft will expect the cliPackageName package 
>  *     to define a binary command named "build-storybook". For the dev server mode,
>  *     Heft will expect to find a binary command named "start-storybook". These commands
>  *     must be declared in the "bin" section of package.json since Heft invokes the script directly.
>  *     The output folder will be specified using the "--output-dir" CLI parameter.
>  *
>  * - "storybook7": Heft looks for a single binary command named "sb". It will be invoked as
>  *    "sb build" for static builds, or "sb dev" for dev server mode.
>  *     The output folder will be specified using the "--output-dir" CLI parameter.
>  */
> "cliCallingConvention": "storybook7",
> ```
> 
> We could go a step further and infer `"storybook6"` versus `"storybook7"` from the **package.json** version, but I think explicitly documenting the behavior might be easier for users to understand when troubleshooting, and it allows other possibilities such as "legacy" e.g. if we needed to reintroduce a more custom setting like `"startupModulePath"` for some reason.

Reply:
I think it makes sense to simplify the options, and the above suggested schema change definitely makes it easier to onboard without having to inspect script paths. Especially if we would be able to provide defaults for storybook 7, one only would need to provide these for storybook 6.

Reply:
updated the PR with the suggestion design proposal, I also added the defaults to improve onboarding into the plugin as in most cases it is not required anymore to provide these options as overrides.